core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,444 @@
+     */
+    private final SimplePublisher<ByteBuffer> publisherToUpstream = new SimplePublisher<>();
+
+    /**
+     * The downstream subscriber that is subscribed to this publisher.
+     */
+    private Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber;
+
+    /**
+     * The amount requested by the downstream subscriber that is still left to fulfill. Updated. when the
+     * {@link Subscription#request(long) request} method is called on the downstream subscriber's subscription. Corresponds to the
+     * number of {@code AsyncResponseTransformer} that will be published to the downstream subscriber.
+     */
+    private final AtomicLong outstandingDemand = new AtomicLong(0);
+
+    /**
+     * This flag stops the current thread from publishing transformers while another thread is already publishing.
+     */
+    private final AtomicBoolean emitting = new AtomicBoolean(false);
+
+    private final Object cancelLock = new Object();
+
+    private SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                 Long maximumBufferSizeInBytes,
+                                 CompletableFuture<ResultT> resultFuture) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(
+            upstreamResponseTransformer, "upstreamResponseTransformer");
+        this.resultFuture = Validate.paramNotNull(
+            resultFuture, "resultFuture");
+        Validate.notNull(maximumBufferSizeInBytes, "maximumBufferSizeInBytes");
+        this.maximumBufferInBytes = Validate.isPositive(
+            maximumBufferSizeInBytes, "maximumBufferSizeInBytes");
+
+        this.resultFuture.whenComplete((r, e) -> {
+            if (e == null) {
+                return;
+            }
+            if (isCancelled.compareAndSet(false, true)) {
+                handleFutureCancel(e);
+            }
+        });
+    }
+
+    /**
+     * @param downstreamSubscriber the {@link Subscriber} to the individual AsyncResponseTransformer
+     */
+    @Override
+    public void subscribe(Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber) {
+        if (downstreamSubscriber == null) {
+            throw new NullPointerException("downstreamSubscriber must not be null");
+        }
+        this.downstreamSubscriber = downstreamSubscriber;
+        downstreamSubscriber.onSubscribe(new DownstreamSubscription());
+    }
+
+    /**
+     * The subscription implementation for the subscriber to this SplittingTransformer.
+     */
+    private final class DownstreamSubscription implements Subscription {
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstreamSubscriber.onError(new IllegalArgumentException("Amount requested must be positive"));
+                return;
+            }
+            if (isCancelled.get()) {
```

Comment:
Do we need to check isCancelled here (it's kind of expensive operation)? It seems doEmit will check it.